src/clj/puppetlabs/puppetserver/certificate_authority.clj

```diff
@@ -1628,42 +1628,60 @@
     (fs/delete csr-path)
     (log/debug (i18n/trs "Removed certificate request for {0} at ''{1}''" subject csr-path))))
 
+(schema/defn prevent-duplicate :- [schema/Int]
+  "Given a list of serials and Puppet's CA CRL, returns vector of serials with
+   any already-revoked serials removed."
+  [serials :- [schema/Int]
+   crl :- X509CRL]
+  (let [crl-revoked-list (.getRevokedCertificates crl)
+        existed-serials (set (map #(.getSerialNumber %) crl-revoked-list))
+        duplicate-serials (set/intersection (set serials) existed-serials)] ; For info output
+        (when (> (count duplicate-serials) 0)
+          (doseq [x duplicate-serials]
+            (log/info (str "Serial " x " existed in CRL.")))) ; Feedback on where to place the info output.
+        (vec (set/difference (set serials) existed-serials))))
+
 (schema/defn revoke-existing-certs!
   "Revoke the subjects' certificates. Note this does not destroy the certificates.
    The certificates will remain in the signed directory despite being revoked."
   [{:keys [signeddir cacert cacrl cakey infra-crl-path
            infra-node-serials-path enable-infra-crl]} :- CaSettings
    subjects :- [schema/Str]]
   (let [[our-full-crl & rest-of-full-chain] (utils/pem->crls cacrl)
-        serials (map #(-> (path-to-cert signeddir %)
-                          (utils/pem->cert)
-                          (utils/get-serial))
-                          subjects)
+        serials (prevent-duplicate (map #(-> (path-to-cert signeddir %)
+                                             (utils/pem->cert)
+                                             (utils/get-serial))
+                                             subjects) our-full-crl)
         serial-count (count serials)
         new-full-crl (utils/revoke-multiple our-full-crl
                                             (utils/pem->private-key cakey)
                                             (.getPublicKey (utils/pem->ca-cert cacert cakey))
-                                            (utils/prevent-duplicate serials our-full-crl))
+                                            serials)
         new-full-chain (cons new-full-crl (vec rest-of-full-chain))]
     (write-crls new-full-chain cacrl)
-    (log/info (i18n/trsn "Revoked 1 certificate: {1}"
-                         "Revoked {0} certificates: {1}"
-                         serial-count
-                         (str/join ", " subjects)))
+    (if (= 0 serial-count)
```

Comment:
IMO, this conditional should wrap the revoking and writing of the chain as well.

Reply:
There's no point in writing the same content to disk again (just creates more load/disk contention), and then you don't need to change jvm-ssl-utils at all.